react_tour/src/exampleState/README.md

```diff
@@ -0,0 +1,204 @@
+
+```JS
+this.state = {
+
+}
+```
+
+You have access to the state object throughout the life of your component. You can update it, render it in your UI, and pass it as props to child components. The state object can be as complex or as simple as you need it to be. Note that you must create a class component by extending React.Component in order to create state like this.
+
+The state object is initialized in the constructor:
+Class example
+
+```JS
+class Car extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+        brand: "Ford"
+        };
+  }
+  render() {
+    return (
+      <div>
+        <h1>My Car</h1>
+      </div>
+    );
+  }
+}
+```
+
+Refer to the state object anywhere in the component by using the ( this.state.propertyname ) syntax
+
+```JS
+class Car extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+        brand: "Ford",
+        color: "red"
+        };
+  }
+  render() {
+    return (
+      <div>
+        <h1>My Car {this.state.brand}</h1>
+        <p> It is a {this.state.color} in color</p>
+      </div>
+    );
+  }
+}
+```
+
+To CHANGE a value in the state object, use the this.setState() method
+When a value in the state object changes, the component will re-render, meaning that the output will change according to the new value(s).
+
+Add a button that will change color;
+
+```JS
+class Car extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      brand: "Ford",
+      color: "red"
+    };
+  }
+  changeColor = () => {
+    this.setState({color: "blue"});
+  }
+  render() {
+    return (
+      <div>
+        <h1>My Car {this.state.brand}</h1>
+        <p> It is a {this.state.color} in color</p>
+        <button
+          type="button"
+          onClick={this.changeColor}
+        >Change color</button>
+      </div>
+    );
+  }
+}
+```
+
+How to declare a new state in react using useState hook? Assuming we use functional components in this whole journey.
+How to update state?
+State accepts any type of data from strings, booleans, arrays, objects, numbers etc, how to update a single key in an object while leaving other keys unaffected? How to update array data( adding or removing an item in an array)?
+Assume array of objects is passes as a state, how to update a single key in a single object of an array without affecting other keys of a particular object and without affecting other objects in the array?
+
+Hooks are a new addition in React 16.8. They let you use state and other React features without writing a class.
+
+//Function example
+
+```JS
+import React, { useState } from 'react';
+
+function Example() {
+  // Declare a new state variable, which we'll call "count"
+  const [count, setCount] = useState(0);
```

Comment:
Can you give an example on how to declare and update state with objects or array datatypes?

Reply:
Some Examples
In each of these example learn to add, edit, delete a single item in the state

```JS
// arrays
const [todos, setTodos] = useState (["Learn hooks", "learn contextapi"])

// arrays of object
const [todos, setTodos] = useState ([
  {
    id: 1,
    name: "Learn hooks!",
  },
   {
    id: 2,
    name: "Learn hooks!",
  },
]}

// objects
const [formData, setFormData] = useState({
  username: "",
  password: ""
})

// objects that contains array
const [person, setPerson] = useState({
  id: 2,
  name: "Baraka",
  age: 28,
  followers: [
     {
        id: 3,
        name: "Winston",
        age: 43
      },
     {
        id: 15,
        name: "Anna",
        age: 22
      },
  ]
})
```